File: Data Structures & Algorithms/generate-parentheses/submission-8.cs
public class Solution {
    public bool Valid(string s) {
        int open = 0;
        foreach (char c in s) {
            open += (c == '(') ? 1 : -1;
            if (open < 0) return false;
        }
        return open == 0;
    }

    public void Dfs(string s, List<string> res, int n) {
        if (s.Length == 2 * n) {
            if (Valid(s)) res.Add(s);
            return;
        }
        Dfs(s + '(', res, n);
        Dfs(s + ')', res, n);
    }

    public List<string> GenerateParenthesis(int n) {
        List<string> res = new List<string>();
        Dfs("", res, n);
        return res;
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-8.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    checkValidString(s) {
        function dfs(i, open) {
            if (open < 0) return false;
            if (i === s.length) return open === 0;

            if (s[i] === '(') {
                return dfs(i + 1, open + 1);
            } else if (s[i] === ')') {
                return dfs(i + 1, open - 1);
            } else {
                return dfs(i + 1, open) || 
                    dfs(i + 1, open + 1) || 
                    dfs(i + 1, open - 1);
            }
        }

        return dfs(0, 0);
    }
}